Olá Neyllor!

Meu nome é Ana Paula Araújo. Fico feliz em rever seu projeto hoje.

Quando vejo um erro pela primeira vez, vou apenas apontá-lo. Vou deixar você encontrá-lo e consertá-lo. Também ao longo do texto farei algumas observações sobre melhorias no código e também farei comentários sobre suas percepções sobre o assunto. Estarei aberta a feedbacks e discussões sobre o tema.

Você encontrará meus comentários abaixo - **por favor, não os mova, modifique ou exclua**.

Você pode encontrar meus comentários em caixas verdes, amarelas ou vermelhas como esta:

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Sucesso. Tudo foi feito corretamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Observações. Algumas recomendações.
</div>

<div class="alert alert-block alert-danger">

<b>Comentário do revisor: </b> <a class="tocSkip"></a>
    
Precisa de correções. O bloqueio requer algumas correções. O trabalho não pode ser aceito com os comentários em vermelho.
</div>

Você pode me responder usando isso:

<div class="alert alert-block alert-info">
<b>Resposta do Aluno: Realizei as devidas correções.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b> Comentário geral do revisor</b> <a class="tocSkip"></a>

Neyllor, obrigada por enviar seu projeto. Seu projeto esta excelente!

A maioria das minhas observações foi em tom de sugestão para ajudá-lo em futuras aplicações, eles não significam que você fez algo errado, mas são coisas que quero destacar que podem ajudá-lo a desenvolver ainda mais suas habilidades. Gostei muito do seu código, você demonstrou um conhecimento amplo no assunto e aplicou alguns metodos que ainda nao foram ensinados no curso! Parabéns.

<ul style="color:red"> 
      <li> Seu trabalho teve apenas um bloco vermelho, que foi referente a analise das hipoteses. Eu fiz alguns observações em amarelo. </li> 

</ul>


Outros pontos são mais elaborados ao decorrer de cada bloco. Cada problema com nosso código é uma chance para aprendermos algo novo.
Boa sorte! :)


    
</div>

<div class="alert alert-block alert-success">
<b> Comentário geral do revisor v.2</b> <a class="tocSkip"></a>

Neyllor obrigada por enviar seu projeto. Fiquei muito feliz com as suas alterações, obrigada pelo tempo dedicado para faze-las. E agora o seu trabalho esta perfeito e você está pronto para a nova sprint! 
    
Eu risquei os meus comentários anteriores nas partes que voce modificou. E meus novos comentário estao marcados da seguinte forma:


<b> Comentário do revisor v.2: </b> 

 Obrigada e Boa sorte ;)

# Análise do risco de inadimplência dos mutuários

Seu projeto é preparar um relatório para a divisão de empréstimos de um banco. Você precisará descobrir se o estado civil de um cliente e o número de filhos têm impacto sobre se ele deixará de pagar um empréstimo. O banco já tem alguns dados sobre a capacidade de crédito dos clientes.

Seu relatório será considerado ao criar uma **pontuação de crédito** de um cliente em potencial. A **contagem de crédito** é usada para avaliar a capacidade de um devedor em potencial de pagar seu empréstimo.


<div class="alert alert-block alert-warning">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Sugiro descrever um pouco mais sobre o projeto e deixar claro qual o objetivo, como tambem a criação de um index. 

    
</div>

## Abra o arquivo de dados e veja as informações gerais.


In [1]:
# Carregando todas as bibliotecas
import pandas as pd
import numpy as np
# Carregue os dados
credit_scoring = pd.read_csv('/datasets/credit_scoring_eng.csv')
credit_scoring.head()

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

<div class="alert alert-block alert-warning">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

O arquivo foi aberto corretamente. Sugiro que a importação das bibliotecas seja feita em uma celula separada.
    
</div>

## Tarefa 1. Exploração de dados

**Descrição dos dados**
- `children` - o número de crianças na família
- `days_employed` - experiência de trabalho em dias
- `dob_years` - idade do cliente em anos
- `education` - educação do cliente
- `education_id` - identificador de educação
- `family_status` - estado civil do cliente
- `family_status_id` - identificador de estado civil
- `gender` - gênero do cliente
- `income_type` - tipo de emprego
- `debt` - havia alguma dívida no pagamento do empréstimo
- `total_income` - renda mensal
- `purpose` - o objetivo de obter um empréstimo


In [2]:
# Vamos ver quantas linhas e colunas nosso conjunto de dados tem
credit_scoring.describe()


children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

In [3]:
# vamos exibir as primeiras 10 linhas
credit_scoring.head(10)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

Ao observar a amostra de dados pode-se verificar que existem valores negativos na coluna "days_employed", e que na coluna "education" não se seguiu um padrão no cadastro da informação, ocasionando da mesma informação esta escrita de diversas formas diferentes.

Além destes fatos, podemos observar valores muito altos na coluna "days_employed" e valores negativos na coluna "children".

<div class="alert alert-block alert-warning">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Boas observações! Você pode destacar outros problemas com os dados, como podemos ver no describe, onde temos valores muito grandes na coluna days_employed, como também o valor minimo da coluna children. 
    
</div>

In [4]:
# Obter informações sobre dados
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Ao buscar por valores ausentes verifica-se que as colunas "days_employed" e "total_income" possuem, em cada uma, um total de 2174 linhas com valores ausentes de um total de 21.525 registros, ou seja, aproximadamente 10,1%.

In [5]:
# Vejamos a tabela filtrada com valores ausentes na primeira coluna com dados ausentes
credit_scoring_filter = credit_scoring.loc[(credit_scoring['days_employed'].isna())]
credit_scoring_filter.head(10)

children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   
65         0            NaN         21  secondary education             1   
67         0            NaN         52    bachelor's degree             0   
72         1            NaN         32    bachelor's degree             0   
82         2            NaN         50    bachelor's degree             0   
83         0            NaN         52  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   
65          unmarried                 4      M       business     0   
67            married                 0      F        retiree     0   
72            married                 0      M  civil servant     0   
82            married                 0      F       employee     0   
83            married                 0      M       employee     0   

    total_income                                   purpose  
12           NaN                         to have a wedding  
26           NaN                                 education  
29           NaN                    building a real estate  
41           NaN                  second-hand car purchase  
55           NaN                         to have a wedding  
65           NaN  transactions with commercial real estate  
67           NaN       purchase of the house for my family  
72           NaN  transactions with commercial real estate  
82           NaN                                   housing  
83           NaN                                   housing

Observa-se que todas as linhas que possuem valores ausentes em "days_employed" também possuem valores ausentes em "total_income" o que pode significar alguma relação entre as duas colunas.


In [6]:
# Vamos aplicar várias condições para filtrar dados e observar o número de linhas na tabela filtrada.
credit_scoring_filter.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 2174 entries, 12 to 21510
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          2174 non-null   int64  
 1   days_employed     0 non-null      float64
 2   dob_years         2174 non-null   int64  
 3   education         2174 non-null   object 
 4   education_id      2174 non-null   int64  
 5   family_status     2174 non-null   object 
 6   family_status_id  2174 non-null   int64  
 7   gender            2174 non-null   object 
 8   income_type       2174 non-null   object 
 9   debt              2174 non-null   int64  
 10  total_income      0 non-null      float64
 11  purpose           2174 non-null   object 
dtypes: float64(2), int64(5), object(5)
memory usage: 220.8+ KB


**Conclusão intermediária**

Observando os dados obtidos através da filtragem de valores ausentes em "days_employed" percebe-se que realmente todas as linhas com valores ausentes nesta coluna, também possuem valores ausentes na coluna "total_income".

Devido a porcentagem de valores ausentes perante a todo o dataset ser de aproximadamente 10,1%, foram feitas analises para se descobrir possiveis relações dos dados desta coluna com outras colunas do dataset.

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Excelente!
    
</div>

In [7]:
# Vamos investigar clientes que não possuem dados sobre as características identificadas e a coluna com os valores ausentes
credit_scoring_nan = credit_scoring[(credit_scoring['days_employed'].isnull()) & (credit_scoring['total_income'].isnull())]
credit_scoring_nan.head()


children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   

    total_income                   purpose  
12           NaN         to have a wedding  
26           NaN                 education  
29           NaN    building a real estate  
41           NaN  second-hand car purchase  
55           NaN         to have a wedding

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Excelente uso dos operadores lógicos para esse filtro!
    
</div>

In [8]:
# Verificar a distribuição pelo sexo no conjuto de dados filtrado
credit_scoring_nan['gender'].value_counts(normalize=True)

F    0.682613
M    0.317387
Name: gender, dtype: float64

In [9]:
# Verificar a distribuição pelo sexo em todo o conjunto de dados
credit_scoring['gender'].value_counts(normalize=True)

F      0.661370
M      0.338583
XNA    0.000046
Name: gender, dtype: float64

In [10]:
# Verificar a distribuição pelo status familiar no conjunto de dados filtrado
credit_scoring_nan['family_status'].value_counts(normalize=True)

married              0.568997
civil partnership    0.203312
unmarried            0.132475
divorced             0.051518
widow / widower      0.043698
Name: family_status, dtype: float64

In [11]:
# Verificar a distribuição pelo status familiar em todo o conjunto de dados
credit_scoring['family_status'].value_counts(normalize=True)

married              0.575145
civil partnership    0.194053
unmarried            0.130685
divorced             0.055517
widow / widower      0.044599
Name: family_status, dtype: float64

In [12]:
# Verificar a distribuição pelo tipo de emprego no conjunto de dados filtrado
credit_scoring_nan['income_type'].value_counts(normalize=True)

employee         0.508280
business         0.233671
retiree          0.189972
civil servant    0.067617
entrepreneur     0.000460
Name: income_type, dtype: float64

In [13]:
# Verificar a distribuição pelo tipo de emprego em todo o conjunto de dados
credit_scoring['income_type'].value_counts(normalize=True)

employee                       0.516562
business                       0.236237
retiree                        0.179141
civil servant                  0.067782
unemployed                     0.000093
entrepreneur                   0.000093
student                        0.000046
paternity / maternity leave    0.000046
Name: income_type, dtype: float64

In [14]:
#Verificar a distribuição pela educação no conjunto de dados filtrado
credit_scoring_nan['education'].value_counts(normalize=True)

secondary education    0.647654
bachelor's degree      0.228151
SECONDARY EDUCATION    0.030819
Secondary Education    0.029899
some college           0.025299
Bachelor's Degree      0.011500
BACHELOR'S DEGREE      0.010580
primary education      0.008740
Some College           0.003220
SOME COLLEGE           0.003220
Primary Education      0.000460
PRIMARY EDUCATION      0.000460
Name: education, dtype: float64

In [15]:
#Verificar a distribuição pela educação em todo conjunto de dados
credit_scoring['education'].value_counts(normalize=True)

secondary education    0.638792
bachelor's degree      0.219187
SECONDARY EDUCATION    0.035865
Secondary Education    0.033031
some college           0.031034
BACHELOR'S DEGREE      0.012729
Bachelor's Degree      0.012451
primary education      0.011614
Some College           0.002184
SOME COLLEGE           0.001347
PRIMARY EDUCATION      0.000790
Primary Education      0.000697
graduate degree        0.000186
GRADUATE DEGREE        0.000046
Graduate Degree        0.000046
Name: education, dtype: float64

In [16]:
#Verificar a distribuição pelo número de filhos no conjunto de dados filtrado
credit_scoring_nan['children'].value_counts(normalize=True)

 0     0.661914
 1     0.218491
 2     0.093836
 3     0.016559
 20    0.004140
 4     0.003220
-1     0.001380
 5     0.000460
Name: children, dtype: float64

In [17]:
#Verificar a distribuição pelo número de filhos em todo conjunto de dados
credit_scoring['children'].value_counts(normalize=True)

 0     0.657329
 1     0.223833
 2     0.095470
 3     0.015331
 20    0.003531
-1     0.002184
 4     0.001905
 5     0.000418
Name: children, dtype: float64

**Possíveis motivos para valores ausentes nos dados**

Após analisar os dados não se observou nenhuma diferença relevante entre os dados dos dois datasets. Por este motivo os valores serão preenchidos da seguinte maneira:

days_employed: A idade dos clientes serão categorizadas por décadas, a partir dela obteremos o valor médio de dias trabalhados para cada categoria de idade, esta média será a utilizada para preeencher os valores ausentes nesta coluna.  

total_income: Será preenchida a partir da média obtida dos dados que se enquadrem na mesma categoria de idade e educação do cliente. 

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Otima analise. É bem interessante fazer analisar essa distribuição de algumas colunas tanto no df original como no df com valores nulos!
    
</div>

## Transformação de dados


In [18]:
# Vamos ver todos os valores na coluna de educação para verificar se e quais grafias precisarão ser corrigidas
credit_scoring['education'].unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

In [19]:
# Corrija os registros, se necessário
credit_scoring['education'] = credit_scoring['education'].str.lower()


In [20]:
# Verificando todos os valores na coluna para ter certeza de que os corrigimos
credit_scoring['education'].unique()

array(["bachelor's degree", 'secondary education', 'some college',
       'primary education', 'graduate degree'], dtype=object)

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Essa padronização ajuda muito na analise. Otimo uso do .str.lower()
    
</div>

In [21]:
# Vamos ver a distribuição de valores na coluna `children`
credit_scoring['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

O dataset apresenta na coluna children dois valores que fogem da normalidade. O primeiro mostra que o valor -1 esta presente em 47 linhas da base de dados, o que corresponde a aproximadamente 0,22% de todo o conjunto de dados. O segundo mostra que o valor 20 esta presente em 76 linhas da base de dados, o que corresponde a aproximadamente 0,35% de todo o conjunto de dados. Logo, juntos representam 0,57% de todos os registros do dataset.
Devido a sua pequena relevancia comparada a base de dados, os mesmos serão removidos da analise. 

In [22]:
# [corrija os dados com base na sua decisão]
credit_scoring = credit_scoring[((credit_scoring['children'] > -1) & (credit_scoring['children'] < 20))]

In [23]:
# Verificar a coluna `children` novamente para ter certeza de que está tudo corrigido
credit_scoring['children'].value_counts()

0    14149
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Concordo com você! Como são poucos dados é interessante deletar. Você pode usar o drop para deletar! 
    
</div>

In [24]:
# Encontre dados problemáticos em `days_employed`, se existirem, e calcule a porcentagem
credit_scoring['days_employed'].describe()

count     19240.000000
mean      63159.820777
std      140928.943329
min      -18388.949901
25%       -2747.235601
50%       -1203.934202
75%        -289.740178
max      401755.400475
Name: days_employed, dtype: float64

In [25]:
cs_days_employed = credit_scoring[(credit_scoring['days_employed'] < 0)]
print("O percentual de linhas com valores negativos em comparação ao total de linhas da tabela é de {:%}".format(len(cs_days_employed)/len(credit_scoring)))

O percentual de linhas com valores negativos em comparação ao total de linhas da tabela é de 73.866928%


Observa-se que a coluna "days_employed" possue uma quantidade grande de dados negativos (73,86%), além de valores extremamente altos. 
Para os valores negativos a abordagem a ser tomada será obter o valor absoluto do mesmo, os transformando em positivos. E para os valores que fujam dos padrões aceitaveis de dias trabalhados, posteriormente será adotada a mesma estrategia a ser utilizada nos valores ausentes.

In [26]:
# Obtendo os valores absolutos
credit_scoring['days_employed'] = abs(credit_scoring['days_employed'])
credit_scoring['days_employed'].describe()

count     19240.000000
mean      67027.691459
std      139130.846446
min          24.141633
25%         927.984311
50%        2195.251592
75%        5556.372075
max      401755.400475
Name: days_employed, dtype: float64

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Excelente uso do `abs()` 
</div>

In [27]:
# Verifique o `dob_years` para valores suspeitos e conte a porcentagem
credit_scoring['dob_years'].describe()

count    21402.000000
mean        43.300206
std         12.579055
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [28]:
filtred_dob_years = credit_scoring[credit_scoring['dob_years'] < 18]
print("O percentual de linhas com valores menores do que 18 anos é de {:%}".format(len(filtred_dob_years)/len(credit_scoring)))

O percentual de linhas com valores menores do que 18 anos é de 0.467246%


Levando-se em consideração que para se conseguir um empréstimo seja necessário que a pessoa seja maior de 18 anos e considerando que o percentual de linhas com valores inferiores a este é menor do 0,5%. Será realizado a exclusão destas linhas do dataframe.

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Otimo! Bem justificado!
    
</div>

In [29]:
# Resolva os problemas na coluna `dob_years`, se existirem
credit_scoring = credit_scoring[credit_scoring['dob_years'] >= 18]

In [30]:
# Verifique o resultado - certifique-se de que está corrigido
credit_scoring['dob_years'].describe()

count    21302.000000
mean        43.503474
std         12.252843
min         19.000000
25%         33.000000
50%         43.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [31]:
# Vamos ver os valores da coluna family_status
credit_scoring['family_status'].unique()


array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

Não foram observados problemas nos dados das colunas "credit_scoring" e "family_status".

<div class="alert alert-block alert-warning">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

<s>Você pode comentar que esta tudo certo com os dados da coluna e que eles não precisam de modificação! 
    
</div>

<div class="alert alert-block alert-success">
<b> Comentário do revisor v.2: </b> <a class="tocSkip"></a>

Obrigada pela modificação! 
    
</div>

In [32]:
# Vamos ver os valores na coluna gender
credit_scoring.groupby('gender')['gender'].count()

gender
F      14083
M       7218
XNA        1
Name: gender, dtype: int64

Ao analisar a coluna gender observa-se que existe um valor que difere de F (Feminino) e M (Masculino) a mesma será removida do dataset.

In [33]:
# Aborde os valores problemáticos, se existirem
credit_scoring = credit_scoring[credit_scoring['gender'] != 'XNA']

In [34]:
# Verifique o resultado - certifique-se de que está corrigido
credit_scoring.groupby('gender')['gender'].count()

gender
F    14083
M     7218
Name: gender, dtype: int64

In [35]:
# Vamos ver os valores na coluna income_type
credit_scoring['income_type'].unique()

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

**Buscando Duplicatas**

In [36]:
# Verificar duplicatas
credit_scoring.duplicated().sum()


71

In [37]:
# Aborde as duplicatas, se existirem
credit_scoring = credit_scoring.drop_duplicates().reset_index(drop = True)

In [38]:
# Última verificação se temos duplicatas
credit_scoring.duplicated().sum()

0

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Excelente!
    
</div>

In [39]:
#Verifique o tamanho do conjunto de dados que você tem agora após suas primeiras manipulações com ele
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21230 entries, 0 to 21229
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21230 non-null  int64  
 1   days_employed     19149 non-null  float64
 2   dob_years         21230 non-null  int64  
 3   education         21230 non-null  object 
 4   education_id      21230 non-null  int64  
 5   family_status     21230 non-null  object 
 6   family_status_id  21230 non-null  int64  
 7   gender            21230 non-null  object 
 8   income_type       21230 non-null  object 
 9   debt              21230 non-null  int64  
 10  total_income      19149 non-null  float64
 11  purpose           21230 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 1.9+ MB


Após finalizado o processo de transformação de dados e remoção das linhas duplicadas o datasets passou a ter 21.230 linhas, de um total original de 21.525 linhas. Uma diferença percentual de aproximadamente 1,37%.


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Mesmo deletando alguns dados, vimos que não comprometeu muito a quantidade de dados. Mas ressalto que é importante a gente analisar esse valor total que esta sendo deletado! Sugiro que não passe de 1% dos dados!
    
</div>

# Trabalhando com valores ausentes

Como observado anteriormente as colunas com valores ausentes são "total_income" e "days_employed". As mesmas serão preenchidas da seguinte maneira: 

days_employed: A idade dos clientes serão categorizadas por décadas, a partir dela obteremos o valor médio de dias trabalhados para cada categoria de idade, esta média será a utilizada para preeencher os valores ausentes nesta coluna.  

total_income: Será preenchida a partir da média da renda dos clientes com a mesma categoria de idade e o mesmo nivel de educação. 

### Restaurar valores ausentes em `total_income`

In [40]:
# Vamos escrever uma função que calcule a categoria de idade
def categoriza_idade(row):
    if (row['dob_years'] < 20):
        return '18-19'
    elif(row['dob_years'] < 30):
        return '20-29'
    elif(row['dob_years'] < 40):
        return '30-39'
    elif(row['dob_years'] < 50):
        return '40-49'
    elif(row['dob_years'] < 60):
        return '50-59'
    elif(row['dob_years'] < 70):
        return '60-69'
    else:
        return '+70'  

In [41]:
#Criando coluna de categoria de idades
credit_scoring['age_category'] = credit_scoring.apply(categoriza_idade, axis=1)

In [42]:
# Verificar como os valores na nova coluna
credit_scoring.head(10)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    6929.865299         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose age_category  
0                purchase of the house        40-49  
1                         car purchase        30-39  
2                purchase of the house        30-39  
3              supplementary education        30-39  
4                    to have a wedding        50-59  
5                purchase of the house        20-29  
6                 housing transactions        40-49  
7                            education        50-59  
8                     having a wedding        30-39  
9  purchase of the house for my family        40-49

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Otima função! Ficou bem interessante essa categorização por decada!
    
</div>

In [43]:
# Crie uma tabela sem valores ausentes e exiba algumas de suas linhas para garantir que ela fique boa
credit_scoring_filter = credit_scoring[pd.notnull(credit_scoring['total_income'])]
credit_scoring_filter.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19149 entries, 0 to 21229
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          19149 non-null  int64  
 1   days_employed     19149 non-null  float64
 2   dob_years         19149 non-null  int64  
 3   education         19149 non-null  object 
 4   education_id      19149 non-null  int64  
 5   family_status     19149 non-null  object 
 6   family_status_id  19149 non-null  int64  
 7   gender            19149 non-null  object 
 8   income_type       19149 non-null  object 
 9   debt              19149 non-null  int64  
 10  total_income      19149 non-null  float64
 11  purpose           19149 non-null  object 
 12  age_category      19149 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.0+ MB


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Gostei do uso do notnull para essa analise
</div>

In [44]:
# Veja os valores médios de renda com base em seus fatores identificados
cs_filter_pivot = credit_scoring_filter.pivot_table(index = ['education','age_category'], values='total_income', aggfunc = 'mean')

In [45]:
# Veja os valores medianos de renda com base em seus fatores identificados
cs_filter_pivot

total_income
education           age_category              
bachelor's degree   +70           26173.068696
                    20-29         29382.496556
                    30-39         34263.111162
                    40-49         35762.892649
                    50-59         33626.170902
                    60-69         30338.012779
graduate degree     30-39         18187.301500
                    40-49         31771.321000
                    50-59         42945.794000
                    60-69         28334.215000
primary education   +70           18892.886000
                    20-29         27695.271520
                    30-39         21979.584515
                    40-49         23618.267667
                    50-59         17797.322623
                    60-69         18710.592883
secondary education +70           19245.043953
                    18-19         16011.592250
                    20-29         23432.406626
                    30-39         25644.051001
                    40-49         26200.754161
                    50-59         24045.961273
                    60-69         21697.500807
some college        +70           13917.989667
                    18-19         18565.702800
                    20-29         25411.042785
                    30-39         32041.530075
                    40-49         33703.486528
                    50-59         27623.492353
                    60-69         29570.940682

A renda muitas vezes é definida pelo nivel de formação e pela experiencia da pessoa. Por este motivo sera utilizado as colunas "education" e "age_category" para definir um valor médio para "total_income".

In [46]:
#  Escreva uma função que usaremos para preencher os valores ausentes
def fill_total_income(row):
    if pd.isna(row['total_income']):
        return cs_filter_pivot['total_income'][row['education']][row['age_category']]
    return row['total_income']

In [47]:
# Aplique em todas as linhas
credit_scoring['total_income'] = credit_scoring.apply(fill_total_income, axis=1)

In [48]:
# Verifique se temos algum erro
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21230 entries, 0 to 21229
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21230 non-null  int64  
 1   days_employed     19149 non-null  float64
 2   dob_years         21230 non-null  int64  
 3   education         21230 non-null  object 
 4   education_id      21230 non-null  int64  
 5   family_status     21230 non-null  object 
 6   family_status_id  21230 non-null  int64  
 7   gender            21230 non-null  object 
 8   income_type       21230 non-null  object 
 9   debt              21230 non-null  int64  
 10  total_income      21230 non-null  float64
 11  purpose           21230 non-null  object 
 12  age_category      21230 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Muito bom. A analise da pivot_table, e a criaçao da função para substituir esses valores nulos ficou otima
</div>

###  Restaurar valores em `days_employed`

[Pense nos parâmetros que podem ajudá-lo a restaurar os valores ausentes nesta coluna. Eventualmente, você desejará descobrir se deve usar valores médios ou medianos para substituir valores ausentes. Você provavelmente fará uma pesquisa semelhante à que fez ao restaurar dados em uma coluna anterior.]

In [49]:
# Distribuição de `days_employed` medianos com base em seus parâmetros identificados
credit_scoring.groupby('age_category')['days_employed'].median()

age_category
+70      361336.993449
18-19       724.492610
20-29      1005.629955
30-39      1601.919871
40-49      2112.694563
50-59      4823.464469
60-69    354932.869424
Name: days_employed, dtype: float64

In [50]:
# Distribuição de `days_employed` médios com base em seus parâmetros identificados
credit_scoring.groupby('age_category')['days_employed'].mean()

age_category
+70      320819.151927
18-19       633.678086
20-29      1969.963146
30-39      4168.767451
40-49     12308.569630
50-59    133225.884547
60-69    283789.958296
Name: days_employed, dtype: float64

Os valores que estão fora do padrão aceitável serão alterados pelo valor médio de dias trabalhados para cada categoria de idade correspondente. 
Adotou-se como valor máximo aceitavel a idade do cliente menos 18, que é o valor em anos que uma pessoa atinge a maiorida legal.
Exemplo:
Se o cliente possui 63 anos, o valor máximo de dias trabalhados deve ser a quantidade de dias que correspondam a 45 anos (60 - 18)

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

É importante analisar a media e mediana!
    
</div>

In [51]:
# Verifique se a função funcionou
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21230 entries, 0 to 21229
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21230 non-null  int64  
 1   days_employed     19149 non-null  float64
 2   dob_years         21230 non-null  int64  
 3   education         21230 non-null  object 
 4   education_id      21230 non-null  int64  
 5   family_status     21230 non-null  object 
 6   family_status_id  21230 non-null  int64  
 7   gender            21230 non-null  object 
 8   income_type       21230 non-null  object 
 9   debt              21230 non-null  int64  
 10  total_income      21230 non-null  float64
 11  purpose           21230 non-null  object 
 12  age_category      21230 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


In [52]:
#Remover outliers da coluna "days_employed"
credit_scoring.loc[(credit_scoring['days_employed']/365) > (credit_scoring['dob_years']-18), 'days_employed'] = float('NaN')

<div class="alert alert-block alert-warning">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Excelente analise, faz sentido essa substituição desses valores discrepantes. É interessante analisar também a porcentagem de dados que estao sendo alterados! 
</div>

In [53]:
# Preencher valores ausentes
credit_scoring['days_employed'] = credit_scoring['days_employed'].fillna(credit_scoring.groupby('age_category')['days_employed'].transform('mean'))

In [54]:
# Verifique as entradas em todas as colunas - certifique-se de corrigir todos os valores ausentes
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21230 entries, 0 to 21229
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21230 non-null  int64  
 1   days_employed     21230 non-null  float64
 2   dob_years         21230 non-null  int64  
 3   education         21230 non-null  object 
 4   education_id      21230 non-null  int64  
 5   family_status     21230 non-null  object 
 6   family_status_id  21230 non-null  int64  
 7   gender            21230 non-null  object 
 8   income_type       21230 non-null  object 
 9   debt              21230 non-null  int64  
 10  total_income      21230 non-null  float64
 11  purpose           21230 non-null  object 
 12  age_category      21230 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Excelent uso do transform
</div>

## Categorização de dados


A coluna "total_income" será categorizada da seguinte maneira:
   - total_income < 10.000: Nivel 1
   - 10.000 < total_income < 20.000: Nivel 2  
   - 20.000 < total_income < 30.000: Nivel 3
   - 30.000 < total_income < 40.000: Nivel 4
   - 40.000 < total_income: Nivel 5     

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Excelente!
</div>

In [55]:
# Exiba os valores dos dados selecionados para categorização
credit_scoring['total_income']


0        40620.102
1        17932.802
2        23341.752
3        42820.568
4        25378.572
           ...    
21225    35966.698
21226    24959.969
21227    14347.610
21228    39054.888
21229    13127.587
Name: total_income, Length: 21230, dtype: float64

In [56]:
credit_scoring['total_income'].quantile(np.linspace(0, 1, 11))

0.0      3306.762000
0.1     12597.063900
0.2     15764.349000
0.3     18555.749500
0.4     21638.419000
0.5     24045.961273
0.6     26200.754161
0.7     29900.771300
0.8     34689.700800
0.9     43174.780800
1.0    362496.645000
Name: total_income, dtype: float64

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Bom uso do linspace, fica melhor para a gente analisar essa distribuição da renda
</div>

In [57]:
# Vamos escrever uma função para categorizar os dados com base em tópicos comunss
def categoriza_renda(row):
    if (row['total_income'] < 10000):
        return 'Nivel 1'
    elif (row['total_income'] < 20000):
        return 'Nivel 2'
    elif (row['total_income'] < 30000):
        return 'Nivel 3'
    elif (row['total_income'] < 40000):
        return 'Nivel 4'
    else:
        return 'Nivel 5'

In [58]:
# Crie uma coluna com as categorias e conte os valores para elas
credit_scoring['nivel_renda'] = credit_scoring.apply(categoriza_renda, axis = 1)
credit_scoring.head()

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    1924.323637         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0    3166.064885         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose age_category nivel_renda  
0    purchase of the house        40-49     Nivel 5  
1             car purchase        30-39     Nivel 2  
2    purchase of the house        30-39     Nivel 3  
3  supplementary education        30-39     Nivel 5  
4        to have a wedding        50-59     Nivel 3

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Categorização foi feita corretamente!
</div>

[Se você decidir categorizar os dados numéricos, também precisará criar as categorias para isso.]

In [59]:
# Examinar todos os dados numéricos em sua coluna selecionada para categorização
credit_scoring['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

In [60]:
# Obter estatísticas resumidas para a coluna
credit_scoring['purpose'].describe()


count                21230
unique                  38
top       wedding ceremony
freq                   785
Name: purpose, dtype: object

[Decida quais intervalos você usará para agrupar e explique por quê.]

In [61]:
# Criar função para categorização em diferentes grupos numéricos com base em intervalos
def categoriza_puporse(row):
    if ((row['purpose'].find('educat') != -1) or (row['purpose'].find('university') != -1)):
        return 'Education'
    if (row['purpose'].find('car') != -1):
        return 'Car'
    elif ((row['purpose'].find('house') != -1) or (row['purpose'].find('real estate') != -1) or
          (row['purpose'].find('property') != -1) or (row['purpose'].find('housing') != -1) or
          (row['purpose'].find('construction') != -1)) :
        return 'Real Estate'
    elif (row['purpose'].find('wedding') != -1):
        return 'Wedding'
    else:
        return 'Other'


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Excelente! A categorizção ficou muita boa. Gostei do uso no find
</div>

In [62]:
# Criar coluna com categorias
credit_scoring['cat_purpose'] = credit_scoring.apply(categoriza_puporse, axis = 1)
credit_scoring.head(10)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    1924.323637         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0    3166.064885         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    1924.323637         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose age_category nivel_renda  cat_purpose  
0                purchase of the house        40-49     Nivel 5  Real Estate  
1                         car purchase        30-39     Nivel 2          Car  
2                purchase of the house        30-39     Nivel 3  Real Estate  
3              supplementary education        30-39     Nivel 5    Education  
4                    to have a wedding        50-59     Nivel 3      Wedding  
5                purchase of the house        20-29     Nivel 5  Real Estate  
6                 housing transactions        40-49     Nivel 4  Real Estate  
7                            education        50-59     Nivel 3    Education  
8                     having a wedding        30-39     Nivel 2      Wedding  
9  purchase of the house for my family        40-49     Nivel 3  Real Estate

In [63]:
# Conte os valores de cada categoria para ver a distribuição
credit_scoring.groupby('cat_purpose')['debt'].count()

cat_purpose
Car             4258
Education       3970
Real Estate    10703
Wedding         2299
Name: debt, dtype: int64

## Verificar as Hipóteses


**Existe uma correlação entre o número de filhos e do pagamento em dia?**

In [65]:
tab_children = credit_scoring.pivot_table(index='children', values='days_employed', columns='debt', aggfunc='count')
# Calcular a taxa de inadimplência com base no número de filhos
tab_children['taxa_inad'] = (tab_children[1])*100/(tab_children[0] + tab_children[1])
tab_children

debt            0       1  taxa_inad
children                            
0         12963.0  1058.0   7.545824
1          4351.0   441.0   9.202838
2          1845.0   194.0   9.514468
3           301.0    27.0   8.231707
4            37.0     4.0   9.756098
5             9.0     NaN        NaN

**Conclusão**

Observa-se que os cliente com 1, 2 e 4 filhos possuem a maior taxa de inadiplência, com 9,2%, 9,51% e 9,75% respectivamente. Clientes com nenhum filho possuem a menor taxa de inadiplência com 7,54% e cliente com 3 filhos possuem uma taxa de 8,23%.


<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

<s>A taxa de inadimplência com base no numero de filhos não foi calculada corretamente! Perceba que quando você divide por `len(credit_scoring)` você esta dividindo pelo tamanho do dataframe e não pelo valor de pessoas que tem 0, 1 ... filhos. Você pode fazer essa divisão usando uma pivot table, que vai ter colunas indicando o numero de pessoas que tem debitos ou nao de acordo com o numero de filhos. Em seguida você pode criar uma nova coluna que faça essa divisão, que seja especifica para quantidade de filhos. 
    Sugiro que essa alteração seja feita nas analises das outras hipoteses também
</div>

<div class="alert alert-block alert-success">
<b> Comentário do revisor v.2: </b> <a class="tocSkip"></a>

Excelente! A taxa de inadimplência relacionada com o numero de filhos foi calculada corretamente
    
</div>

**Existe uma correlação entre o status familiar e o pagamento em dia?**

In [66]:
# Verifique os dados de status da família e do pagamento em dia
tab_family_status = credit_scoring.pivot_table(index='family_status', values='days_employed', columns='debt', aggfunc='count')
# Calcular a taxa padrão com base no status da família
tab_family_status['taxa_inad'] = (tab_family_status[1])*100/(tab_family_status[0] + tab_family_status[1])
tab_family_status

debt                   0    1  taxa_inad
family_status                           
civil partnership   3729  383   9.314202
divorced            1095   84   7.124682
married            11290  923   7.557521
unmarried           2508  272   9.784173
widow / widower      884   62   6.553911

**Conclusão**
 Observa-se que, com relação ao status familiar, os clientes unmarried possuem a maior taxa de inadiplência com 9,78%%, seguido pelos clientes com status "civil partnership" com 9,31%. Os clientes em relacionamentos abertos possuem a menor taxa com 6,55%.

<div class="alert alert-block alert-success">
<b> Comentário do revisor v.2: </b> <a class="tocSkip"></a>

Excelente! A taxa de inadimplência relacionada com o status familiar foi calculada corretamente
    
</div>

**Existe uma correlação entre o nível de renda e o pagamento em dia?**

In [67]:
# Verifique os dados do nível de renda e do pagamento em dia
tab_nivel_renda = credit_scoring.pivot_table(index='nivel_renda', values='days_employed', columns='debt', aggfunc='count')
# Calcular a taxa de inadimplência com base no nível de renda
tab_nivel_renda['taxa_inad'] = (tab_nivel_renda[1])*100/(tab_nivel_renda[0] + tab_nivel_renda[1])
tab_nivel_renda

debt            0    1  taxa_inad
nivel_renda                      
Nivel 1       858   58   6.331878
Nivel 2      5852  548   8.562500
Nivel 3      6925  664   8.749506
Nivel 4      3281  261   7.368718
Nivel 5      2590  193   6.934962

**Conclusão**

Pessoas enquadradas no nivel de renda 2 e 3 possuem a maior taxa de inadiplência com aproximadamente 8,56% e 8,74%, respectivamente, enquanto pessoas do nivel 1 possuem a menor taxa com 6.33%.

<div class="alert alert-block alert-success">
<b> Comentário do revisor v.2: </b> <a class="tocSkip"></a>

Excelente! A taxa de inadimplência relacionada com o nivel de renda foi calculada corretamente
    
</div>

**Como a finalidade do crédito afeta a taxa de inadimplência?**

In [68]:
# Confira os percentuais de inadimplência para cada finalidade de crédito e analise-os
tab_cat_purpose = credit_scoring.pivot_table(index='cat_purpose', values='days_employed', columns='debt', aggfunc='count')
# Calcular a taxa de inadimplência com base na finalidade de crédito
tab_cat_purpose['taxa_inad'] = (tab_cat_purpose[1])*100/(tab_cat_purpose[0] + tab_cat_purpose[1])
tab_cat_purpose

debt            0    1  taxa_inad
cat_purpose                      
Car          3861  397   9.323626
Education    3601  369   9.294710
Real Estate  9926  777   7.259647
Wedding      2118  181   7.872988

**Conclusão**

Quanto a finalidade de crédito, pessoas que utilizam o crédito para realizar a compra de carros ou para educação possuem uma taxa de inadiplência de aproximadamente 9,3%. Com a finalidade de realizar um casamente a taxa é de 7,87% e para comprar algum tipo de imóvel a finalidade é de aproximandamente 7,26%.


<div class="alert alert-block alert-success">
<b> Comentário do revisor v.2: </b> <a class="tocSkip"></a>

Excelente! A taxa de inadimplência relacionada com a finalidade foi calculada corretamente
    
</div>

# Conclusão Geral 

Foram realizados os estudos para responder as seguintes questões:

1: Existe alguma relação entre ter filhos e pagar o empréstimo em dia?

2: Existe alguma relação entre o estado civil e o pagamento do empréstimo?

3: Existe relação entre o nível de renda e o pagemento no prazo?

4: Como as diferentes finalidades do empréstimo afetam o pagamento pontual do mesmo?

Após as analises dos dados chegou-se as seguintes respostas:

1: Observou-se que clientes sem filhos possuem a menor taxa de inadiplência, seguidos por clientes com 3, 1, 2 e 4 filhos.

2: Observou-se que com relação ao status familiar, os clientes solteiros possuem a maior taxa de inadiplência com aproximadamente 9,78%, seguidos pelos clientes em um relacionamento estavel com aproximandamente 9,31%. Os clientes em relacionamentos abertos possuem a menor taxa com 6,559%, os divorciados possuem uma taxa de 7,12%, os casados de 7,55%.

3: Clientes com as rendas de nivel 1 e nivel 5 possuem menores taxas de inadiplência, aproximadamente 6,33% e 6,93% respectivamente. enquanto clientes com renda entre 10.000 e 30.000 possuem as maiores taxas de inadiplência.

4: Clientes que utilizam o emprestimo para compra ou reforma de algum tipo de imóvel possuem as menores taxas de inadiplência, 7,26% aproximadamente. Seguidas de compra de casamaento com 7,87%. Enquanto as taxas mais altas são para empréstimos utilizados para educação, 9,29%, e aquisição de carros com aproximadamente 9,32%.




<div class="alert alert-block alert-success">
<b> Comentário do revisor v.2: </b> <a class="tocSkip"></a>

Excelente!
    
</div>